rfcs/0009-structured-modules.md

```diff
@@ -0,0 +1,190 @@
+use bar::fun4;
+
+pub fn fun2() -> u32 {
+   fun3() + fun4()
+}
+```
+
+* `src/bar/foo.sw`
+```sway
+library;
+
+pub fn fun3() -> u32 {
+   3
+}
+```
+
+* `src/bar/bar.sw`
+```sway
+library;
+
+pub fn fun4() -> u32 {
+   4
+}
+```
+
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation requirements of this change are deliberately small and
+straightforward: introduce `mod` statements to replace `dep` statements, remove
+the need for an ident argument to `library`, deduce the library name from
+the source folder structure, add checks to ensure the structure isn't ambiguous.
+
+The interactions with other features are minimal.
+However we might be able to refactor some compiler behavior after establishing
+that dependency and library names are no longer different.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+This is a breaking change and will require changing all sway module imports and
+restructuring projects that use the existing permissive scheme in creative
+ways, but this is both rare (it seems that most follow Rust like conventions
+already in practice) and will result in better code structure and less
+boilerplate.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could also follow the existing Rust convention and allow both `submodule.sw`
+and `submodule/mod.sw` or only allow the latter instead of the former as this
+proposal does. But Rust's allowance of both is an artifact of backwards
+compatibility rather than a deliberate design choice. It is likely best to
+settle on and enforce a single convention.
+
+If we do not enforce any module structure and hold to the current behavior,
+the issues and confusion will remain and it will become harder to change in the
+future.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Rust's module structure is the prior art we are taking cues from, but there is
+also a [previous, now abandoned,
+RFC](https://github.com/FuelLabs/sway-rfcs/pull/8) that sought similar but more
+wide changes to the way paths and module imports work in Sway.
+
+The present RFC tries to make the change as small and painless as possible so
+that we may build upon it later.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+The elimination of file type headers such as `library;` or `contract;` when
+they aren't needed is not considered here. Neither are wider issues about the
+structure and type of package exports. But clarifying the module structure will
+definitely give us a more stable base to consider these.
+
+We also do not consider the visibility of modules here and reproduce the
+public by default scheme of `dep`.
+
+# Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+We could consider eliminating module type headers like `library;`, `contract;`
+and `predicate;` from the source altogether and defining the exports in the
+`Forc.toml` or implicitly through usage.
```

Comment:
Huge fan of eliminating these from src files in favour of assuming `library` by default, and declaring the top-level program type in the forc manifest. Currently forc basically has to do a parse of the file just to check the program type (it's not safe to just grab the first line as a user might have combinations of whitespace, comments and now module-level comments preceding the declaration). But yes, makes sense for this to be addressed as follow-up.